### ISSUE TITLE
Allow repository rules to report env vars they're sensitive to later in the process

### ISSUE THREAD

**Issue by illicitonion** [bazelbuild/bazel#19511]
### Description of the feature request:

Right now, if a repository rule is sensitive to an env var, it needs to declare it up-front on the `repository_rule` definition.

This means that a repository rule can't safely accept the name of an env var as an attribute, and read that env var as part of its execution, as it can't know in advance what to tell Bazel when it needs to be invalidated.

Ideally there would be some way for a repository rule to tell Bazel that it's sensitive to a particular env var at runtime. Some options I can imagine:
1. Returning a provider-like object from the implementation function which has a field for env vars.
2. A function that can be called on `repository_ctx`.
3. If Bazel were being fancy, it could automatically detect env var reads by wrapping `repository_ctx.os.environ` to look for accesses.

### Which category does this issue belong to?

External Dependency

### What underlying problem are you trying to solve with this feature?

I have a repository rule which is responsible for looking at the environment to work out information that will be used when stamping artifacts with version numbers. Currently I need to generate this `repository_rule` call in a function in order to capture the env var name and set the `environ` argument, but discussion https://github.com/bazelbuild/bazel/issues/19459 suggests this is undesirable.

**Comment by beasleyr-vmw** [comment]
@meteorcloudy Looking at the "help wanted" label, how could a volunteer get started on this one? Are any Googlers available to steer folks?

As a naive outsider, illicitonion's 2nd suggestion seems straightforward to implement, since the semantics could be similar to RepositoryFunction's handling of `rctx.path(label_type)`.

```python
# foo.bzl
def _derive_env(s):
    return s.upper() + "_ROOT"

def _impl(rctx):
    # equivalent to rctx.os.environ.get("MYNAME_ROOT"), but adds
    # ENV:MYNAME_ROOT key to RepositoryFunction marker
    value = rctx.env(_derive_env(rctx.name))

myreporule = repository_rule(_impl, attrs = {...})

# WORKSPACE
myreporule(name = "myname")
```

- Would need to make sure things play nice with `--repo_env`.

**Comment by Wyverald** [comment]
fixed by #20787.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.1.0 RC1](https://releases.bazel.build/7.1.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
